src/utils/generator.js

```diff
@@ -136,8 +137,12 @@ class TemporalBuildNodeEntry extends GeneratorEntry {
   buildNode: void | GeneratorBuildNodeFunction;
   dependencies: void | Array<Generator>;
   isPure: void | boolean;
+  purityCheck: void | ((callbacks: VisitEntryCallbacks, declared: void | Value, args: Array<Value>) => boolean);
 
   visit(callbacks: VisitEntryCallbacks, containingGenerator: Generator): boolean {
+    if (this.purityCheck) {
+      this.isPure = this.purityCheck(callbacks, this.declared, this.args);
```

Comment:
It's unexpected to me that the first visit mutates `this.isPure`.

Reply:
It's mutated so that in the serialization phase, we know it's pure. We can't do the same purity check here, as that's only something that occurs in the visitor phase. I guess we could make another flag?